ruma-api-macros/src/derive_outgoing.rs

```diff
@@ -0,0 +1,184 @@
+use std::mem;
+
+use proc_macro2::{Ident, Span, TokenStream};
+use quote::{format_ident, quote};
+use syn::{
+    parse_quote,
+    punctuated::{Pair, Punctuated},
+    AngleBracketedGenericArguments, Attribute, Data, DeriveInput, Field, Fields, GenericArgument,
+    GenericParam, Generics, ImplGenerics, Path, PathArguments, Token, Type, TypeGenerics, TypePath,
+    TypeReference, TypeSlice,
+};
+
+enum StructKind {
+    Struct,
+    Tuple,
+}
+
+pub fn expand_derive_outgoing(input: DeriveInput) -> syn::Result<TokenStream> {
+    let derive_deserialize = if no_deserialize_in_attrs(&input.attrs) {
+        TokenStream::new()
+    } else {
+        quote!(::ruma_api::exports::serde::Deserialize)
+    };
+
+    let (mut fields, struct_kind): (Vec<_>, _) = match input.data {
+        Data::Enum(_) | Data::Union(_) => {
+            panic!("#[derive(Outgoing)] is only supported for structs")
+        }
+        Data::Struct(s) => match s.fields {
+            Fields::Named(fs) => {
+                (fs.named.into_pairs().map(Pair::into_value).collect(), StructKind::Struct)
```

Comment:
I think this is just `into_iter`:

```suggestion
                (fs.named.into_iter().collect(), StructKind::Struct)
```